Caw tunnel version (#2)
ekr/draft-ietf-tls-esni: draft-ietf-tls-esni.md

backend server, which is the true origin server. The provider does
not have access to the plaintext of the connection.

## ClientHello Encryption

> SNI encryption works by encrypting the entire ClientHello, including

Author: ECHO works?

# Encrypted ClientHello Configuration {#echo-configuration}

ClientHello encryption configuration information is conveyed with the
following ECHOConfigs structure.

~~~~
    opaque HpkePublicKey<1..2^16-1>;
    uint16 HkpeKemId; // Defined in I-D.irtf-cfrg-hpke

    struct {
        opaque public_name<1..2^16-1>;

        HpkePublicKey public_key;
        HkpeKemId kem_id;
        CipherSuite cipher_suites<2..2^16-2>;

        uint16 maximum_name_length;
        Extension extensions<0..2^16-1>;
    } ECHOConfigContents;

    struct {
        uint16 version;
        uint16 length;
        select (ECHOConfig.version) {
          case 0xff03: ECHOConfigContents;
        }
    } ECHOConfig;

    ECHOConfig ECHOConfigs<1..2^16-1>;
~~~~

The ECHOConfigs structure contains one or more ECHOConfig structures in
decreasing order of preference. This allows a server to support multiple
versions of ECHO and multiple sets of ECHO parameters.

The ECHOConfig structure contains the following fields:

version
: The version of the structure. For this specification, that value
SHALL be 0xff03. Clients MUST ignore any ECHOConfig structure with a
version they do not understand.

> [[NOTE: This means that the RFC will presumably have a nonzero value.]]

Author: Why will it have a non-zero value? I see this is old, but I don't see why,

Reviewer: TBH I don't remember why we have this here. Do we want to remove it? I'd be fine with that.

contents
: An opaque byte string whose contents depend on the version of the structure.
For this specification, the contents are an ECHOConfigContents structure.

The ECHOConfigContents structure contains the following fields:

public_name
: The non-empty name of the entity trusted to update these encryption keys.
This is used to repair misconfigurations, as described in
{{handle-server-response}}.

public_key
: The HPKE {{!I-D.irtf-cfrg-hpke}} public key which can be used by the client to
encrypt the ClientHello.

kem_id
: The HPKE {{!I-D.irtf-cfrg-hpke}} KEM identifier corresponding to public_key.
Clients MUST ignore any ECHOConfig structure with a key using a KEM they do not support.

cipher_suites
: The list of cipher suites which can be used by the client to encrypt the ClientHello.
See {{hpke-map}} for information on how a CipherSuite maps to corresponding

> HPKE algorithm identifiers.

Author: ```suggestion
: The list of TLS cipher suites which can be used by the client to encrypt the ClientHello.
See {{hpke-map}} for information on how a cipher suite maps to corresponding
HPKE algorithm identifiers.
```

of the structure.  This hash is computed using the hash function
associated with `suite`.

enc
: The HPKE encapsulated key, used by servers to decrypt the corresponding
encrypted_ch field.

encrypted_ch
: The serialized and encrypted ClientHelloInner structure, AEAD-encrypted using
cipher suite "suite" and the key generated as described below.
{:br}

If the server accepts ECHO, it does not send this extension.
If it rejects ECHO, then it sends the following structure in
EncryptedExtensions:

~~~
   struct {
       ECHOConfigs retry_configs;
   } ServerEncryptedCH;
~~~

retry_configs
: An ESNIConfigs structure containing one or more ECHOConfig structures in
decreasing order of preference that the client should use on subsequent
connections to encrypt the ClientHelloInner structure.

This protocol also defines the "echo_required" alert, which is sent by the
client when it offered an "encrypted_server_name" extension which was not
accepted by the server.

# The "echo_nonce" extension {#echo-nonce}

When using ECHO, the client MUST also add an extension of type
"echo_nonce" to the ClientHelloInner (but not to the outer

> ClientHello). This extension is defined as follows:

Author: ```suggestion
When using ECHO, the client MUST also add an extension of type
"echo_nonce" to the ClientHelloInner (but not to the outer
ClientHello). This nonce ensures that the server's encrypted
Certificate can only be read by the entity which sent this
ClientHello. [[TODO: Describe HRR cut-and-paste 1 in
Security Considerations.]]
This extension is defined as follows:
```

a very large overall size. One particularly pathological example
is "key_share" with post-quantum algorithms. In order to reduce
the impact of duplicated extensions, the client may use the
"outer_extensions" extension.

~~~
   enum {
       outer_extension(TBD), (65535)
   } ExtensionType;

   struct {
       ExtensionType outer_extensions<2..254>;
       uint8 hash<32..255>;
   } OuterExtensions;
~~~~

OuterExtensions MUST only be used in ClientHelloInner. It consists
of one or more ExtensionType values, each of which reference an
extension in ClientHelloOuter, and a digest of the complete
ClientHelloInner.

When sending ClientHello, the client first computes ClientHelloInner,
including any PSK binders, and then MAY substitute extensions which
it knows will be duplicated in ClientHelloOuter. To do so, the client
computes a hash H of the entire ClientHelloInner message with the same
hash as for the KDF used to encrypt ClienHelloInner. Then, the client
removes and and replaces extensions from ClientHelloInner with a single
"outer_extensions" extension. The list of outer_extensions include those
which were removed from ClientHelloInner, in the order in which they were

> removed. The hash contains full ClientHelloInner hash H computed above.

Author: ```suggestion
removed. The hash contains the full ClientHelloInner hash H computed above.
```

This process is reversed by client-facing servers upon receipt. Specifically,
the server replaces the "outer_extensions" with extensions contained in
ClientHelloOuter. The server then computes a hash H' of the reconstructed

> ClientHelloInner. If H' does not equal OuterExtensions.hash, the server aborts
> the connection with an "illegal_parameter" alert.

Author: ```suggestion
ClientHelloInner. If H' does not equal OuterExtensions.hash, the server MUST abort
the connection with an "illegal_parameter" alert.
```

that ECHOConfig value and MAY attempt to use another value provided by
the server. The client MUST NOT send ECHO using HPKE algorithms not advertised
by the server.

Given a compatible ECHOConfig with fields public_key and kem_id, carrying the HpkePublicKey
and KEM identifier corresponding to the server, clients compute an HPKE encryption
context as follows:

~~~
pkR = HPKE.KEM.Unmarshal(ECHOConfig.public_key)
enc, context = SetupBaseS(pkR, "tls13-echo")
echo_nonce = context.Export("tls13-echo-nonce", 16)
echo_hrr_key = context.Export("tls13-echo-hrr-key", 16)
~~~

Note that the HPKE algorithm identifiers are those which match the client's
chosen CipherSuite, according to {{hpke-map}}. The client MAY replace any large,
duplicated extensions in ClientHelloInner with the corresponding "outer_extensions"
extension, as described in {{outer-extensions}}.

The client then generates a ClientHelloInner value. In addition to the normal
values, ClientHelloInner MUST also contain:

 - an "echo_nonce" extension
 - TLS padding {{!RFC7685}}

Padding SHOULD be P = L - D bytes, where

- L = ECHOConfig.maximum_name_length, rounded up to the nearest multiple of 16

> - D = len(dns_name), where dns_nae is the DNS name in the ClientHelloInner "server_name" extension

Author: ```suggestion
- D = len(dns_name), where dns_name is the DNS name in the ClientHelloInner "server_name" extension
```

Reviewer: *facepalm*

implemented, for instance, by reporting a failed connection with a dedicated
error code.

### HelloRetryRequest

If the server sends a HelloRetryRequest in response to the ClientHello,
the client sends a second updated ClientHello per the rules in {{RFC8446}}.
However, at this point, the client does not know whether the server processed
ClientHelloOuter or ClientHelloInner, and MUST regenerate both values to
be acceptable. Note: if the inner and outer ClientHellos use different groups
for their key shares or differ in some other way, then the HelloRetryRequest may
actually be invalid for one or the other ClientHello. If the inner ClientHello
is unaffected by this retry, then the client only changes the outer ClientHello.
In contrast, if the outer ClientHello is unaffected by this retry, then the client

> changes the inner ClientHello and recomputes any fields necessary in the outer

Author: This seems like an unwise optimization. What's the rationale here?

Also, won't it be impossible because you'll get a different echo_nonce?

Reviewer: By optimization do you mean it'd be simpler to just re-generate both? I think the intent was to stick with the HRR model and change as little as possible. But I'd be fine with regenerating both inner and outer CHs. That would definitely be simpler to implementer, IMO.

ClientHello ("encrypted_client_hello" extension contents, PSK binders, etc.)

Otherwise, the usual rules for HelloRetryRequest processing apply.

Clients bind encryption of the second ClientHelloInner to encryption of the first
ClientHelloInner via the derived echo_hrr_key by modifying HPKE setup as follows:

~~~
pkR = HPKE.KEM.Unmarshal(ECHOConfig.public_key)

> enc, context = SetupPSKS(pkR, "tls13-echo-hrr", echo_hrr_key, "")

Author: See my previous comments about ctx rather than PSK. Maybe an open issue marker here?

Reviewer: I went back and double checked the PSK variant of HPKE, and `info` is assumed to be known to the attacker in our analysis. Here that's not the case, so I *think* PSK is the right place for it. But agreed, let's drop a TODO to follow up with Karthik et al.

echo_nonce = context.Export("tls13-echo-hrr-nonce", 16)
~~~

Clients then encrypt the second ClientHelloInner using this new HPKE context.
In doing so, the encrypted value is also authenticated by echo_hrr_key.
Client-facing servers perform the corresponding process when decrypting second
ClientHelloInner messages. In particular, upon receipt of a second ClientHello
message with a ClientEncryptedCH value, servers setup their HPKE context and
decrypt ClientEncryptedCH as follows:

~~~
context = SetupPSKR(ClientEncryptedCH.enc, skR, "tls13-echo-hrr", echo_hrr_key, "")
ClientHelloInner = context.Open("", ClientEncryptedCH.encrypted_ch)
echo_nonce = context.Export("tls13-echo-hrr-nonce", 16)
~~~

[[OPEN ISSUE: This, along with trial decryption is
pretty gross. It would just be a lot easier if we were willing to
have the server indicate whether ECHO had been accepted or not.
Given that the server is supposed to only reject ECHO when it doesn't
know the key, and this is easy to probe for, can we just instead

> have an extension to indicate what has happened.]]

Author: Note that this would allow us to do ESHO.

Depending on whether the client is configured to accept the proxy's certificate
as authoritative for the public name, this may trigger the retry logic described
in {{handle-server-response}} or result in a connection failure. A proxy which
is not authoritative for the public name cannot forge a signal to disable ECHO.

A non-conformant MITM proxy which instead forwards the ECHO extension,
substituting its own KeyShare value, will result in
the client-facing server recognizing the key, but failing to decrypt
the SNI. This causes a hard failure. Clients SHOULD NOT attempt to repair the
connection in this case.

# TLS and HPKE CipherSuite Mapping {#hpke-map}

Per {{RFC8446}, TLS ciphersuites define an AEAD and hash algorithm. In contrast,

> HPKE composes AEAD algorithms and key derivation functions. The table below lists

Author: ```suggestion
HPKE defines separate AEAD algorithms and key derivation functions. The table below lists
```
